src/analysis/processing/qgsalgorithmnearestneighbouranalysis.cpp

```diff
@@ -0,0 +1,155 @@
+                      "Output is generated as an html file with the computed statistical values." );
+}
+
+QString QgsNearestNeighbourAnalysisAlgorithm::svgIconPath() const
+{
+  return QgsApplication::iconPath( QStringLiteral( "/algorithms/mAlgorithmNearestNeighbour.svg" ) );
+}
+
+QIcon QgsNearestNeighbourAnalysisAlgorithm::icon() const
+{
+  return QgsApplication::getThemeIcon( QStringLiteral( "/algorithms/mAlgorithmNearestNeighbour.svg" ) );
+}
+
+QgsNearestNeighbourAnalysisAlgorithm *QgsNearestNeighbourAnalysisAlgorithm::createInstance() const
+{
+  return new QgsNearestNeighbourAnalysisAlgorithm();
+}
+
+void QgsNearestNeighbourAnalysisAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  addParameter( new QgsProcessingParameterFileDestination( QStringLiteral( "OUTPUT_HTML_FILE" ), QObject::tr( "Nearest neighbour" ),
+                QObject::tr( "HTML files (*.html *.HTML)" ), QVariant(), true ) );
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "OBSERVED_MD" ), QObject::tr( "Observed mean distance" ) ) );
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "EXPECTED_MD" ), QObject::tr( "Expected mean distance" ) ) );
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "NN_INDEX" ), QObject::tr( "Nearest neighbour index" ) ) );
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "POINT_COUNT" ), QObject::tr( "Number of points" ) ) );
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "Z_SCORE" ), QObject::tr( "Z-score" ) ) );
+}
+
+QVariantMap QgsNearestNeighbourAnalysisAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsProcessingFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  QString outputFile = parameterAsFileOutput( parameters, QStringLiteral( "OUTPUT_HTML_FILE" ), context );
+
+  QgsSpatialIndex spatialIndex( *source, feedback );
+  QgsDistanceArea da;
+  da.setSourceCrs( source->sourceCrs(), context.transformContext() );
+  da.setEllipsoid( context.project()->ellipsoid() );
+
+  double step = source->featureCount() ? 100.0 / source->featureCount() : 1;
+  QgsFeatureIterator it = source->getFeatures( QgsFeatureRequest().setSubsetOfAttributes( QList< int >() ) );
+
+  QgsFeatureRequest request;
+  QgsFeature neighbour;
+  double sumDist = 0.0;
+  double area = source->sourceExtent().width() * source->sourceExtent().height();
+
+  int i = 0;
+  QgsFeature f;
+  while ( it.nextFeature( f ) )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    QgsFeatureId neighbourId = spatialIndex.nearestNeighbor( f.geometry().asPoint(), 2 ).at( 1 );
```

Comment:
Hm, we can potentially give this algorithm a huge speed boost -- if we use the FlagStoreFeatureGeometries during the construction of the spatial index, then the geometries will be stored inside the algorithm and can be later retrieved using QgsSpatialIndex::geometry( ... ) . This would avoid the need for all the extra single-feature requests.

Reply:
Should be fixed now